Add unsplash image to body background using axios to get random image

moltenkaizen opened the pull request.

Commit:
Add unsplash image to body background using axios to get random image

angelocordon commented on index.html:
```diff
@@ -3,9 +3,34 @@
   <head>
     <meta charset="utf-8">
     <title>super-fantastic-dashboard</title>
+    <script src="https://unpkg.com/axios/dist/axios.min.js"></script>
```
Let's move all CDN scripts towards the bottom of the page so that it's non-blocking.

moltenkaizen replied:
Excellent, ok

angelocordon commented on index.html:
```diff
@@ -3,9 +3,34 @@
   <body>
     <div id="app"></div>
     <script src="/dist/build.js"></script>
+    <script>
```
Curious what Axios is doing here for us - what is the GET url that we're constructing? 

Can this be done using a simple AJAX request and assign the response to a variable and then passing that variable to a function? While it makes it easier - I just want us to think about tradeoffs - and having unnecessary dependencies. Considering this is the one and only place we're making a 3rd party request, it may be more worth it for now to build out our own _if_ it's simple enough. Happy to be wrong also.

moltenkaizen replied:
I'm open to re-writing with a vanilla ajax request. I've not done that before, but the info is online. Good point, we're only using axios for this one very small thing. I'll take a look at doing that.

angelocordon commented on index.html:
```diff
@@ -3,9 +3,34 @@
+          // console.log(response);
+          document.body.style.backgroundImage = "url(" + response.data.urls.regular + ")";
+        }).catch(function (error) {
+          console.log(error);
```
Over here we'd want a function that picks a random color in case Unsplash doesn't load. Let's source these hex codes into an array and pick random ones on reload. https://flatuicolors.com/ 

Just don't pick the light gray ones since our font color will be white, but I think the rest should work 👍

moltenkaizen replied:
Ok, I'll take a look at doing that also.

angelocordon commented on index.html:
```diff
@@ -3,9 +3,34 @@
   <body>
     <div id="app"></div>
     <script src="/dist/build.js"></script>
+    <script>
```
We'd also want to separate this script to its own file - like background_component.js or something and import it under main.js.

moltenkaizen replied:
Sure thing. Will do. Based on this: https://vuejs-templates.github.io/webpack/static.html I think we'd need to start a static folder. Then we could have css js image or whatever we need under there.

angelocordon reviewed (changes requested):
Great start @moltenkaizen!!

Commit:
Move javascript from index.html to static/js/background_image_loader.js. Move script tags for axios and background_image_loader.js to the bottom of body.

Commit:
Add random fallback background-color to body if the axios promise fails. Color palette from https://flatuicolors.com/ Removed background-color CSS from index.html since the fallback color will now be set in JS

Commit:
Remove axios dependency, instead using fetch.

angelocordon commented on static/js/background_image_loader.js:
```diff
@@ -1,37 +1,50 @@
-        '#7f8c8d' // Asbestos
-    ];
+(function () {
+  "use strict";
```
Use singe quotes here - to keep in consistency among other JS files

angelocordon commented on index.html:
```diff
@@ -3,9 +3,18 @@
   <body>
     <div id="app"></div>
     <script src="/dist/build.js"></script>
+    <script src="/static/js/background_image_loader.js"></script>
```
Can we import this in main.js instead?

jacobarriola replied:
@moltenkaizen I agree. Feels a bit uneasy having this outside of the Vue scope. Perhaps we can even try to make a component out of this or add it to the `App.vue` file. We could probably add these styles to the `#app` ID instead of `<body>`.

Then we could just add the functionality to the vue file instead of a function.

moltenkaizen replied:
That was my original plan but the idea to go this route was mentioned. I'm open to moving this functionality into Vue for sure.

moltenkaizen replied:
@jacobarriola @angelocordon just pushed a new commit with an Unsplash-Background.vue component imported into App.vue. We could also move all the functionality directly inside App.vue.

angelocordon commented on static/js/background_image_loader.js:
```diff
@@ -0,0 +1,50 @@
+    })
+    .catch((error) => {
+      console.log(error);
+      // On error, set the background color to randomly selected color
```
Notes should be more about the `why` instead of the `what`. 

Your code should be written well enough that you don't need to write a note to describe what it's doing. It's okay if you write notes for yourself while you're working on a feature, but make sure it gets cleaned up prior to a PR.

angelocordon commented on static/js/background_image_loader.js:
```diff
@@ -0,0 +1,50 @@
+  fetch(request)
+    .then(response => response.json())
+    .then((response) => {
+      // console.log(response);
```
We don't want `console.log` statements in production.

angelocordon commented on static/js/background_image_loader.js:
```diff
@@ -0,0 +1,50 @@
+  "use strict";
+
+  // Define palette of fallback colors
+  const COLORS = [
```
I think it's okay to clump these hexes together instead of listing it. I also don't think we need to note what the colors are, as long as it's pulling a different color when reloading and when the image doesn't load.

Commit:
Clean up background_image_loader.js. Consistent use of single quotes in js. Removed comments and console.log() statements.

Commit:
Move unsplash background functionality from index.html and background_image_loader.js to a component Unsplash-Background.vue imported into App.vue

jacobarriola committed:
Remove unnecessary template tag

Linter was throwing warnings

jacobarriola committed:
Move image function into a Vue instance

Keeps thing within Vue conventions, plus data is available to use in Vue devtools.

jacobarriola commented:
@moltenkaizen Heyo! Check out the changes I made to your branch. I moved your function into a Vue module/instance. That way some of the data is available to us for debugging and it _feels_ for Vue-ish. Let me know what you think or if I need to go over some things (assumptions, etc) :)

jacobarriola committed:
Merge branch 'master' into feature/background-unsplash

# Conflicts:
#	index.html
